File: 32.R
# Pandigital products

get_products <- function(low_multiplicand, high_multiplicand, low_multiplier,
                         high_multiplier) {

    multiplicand <- rep(low_multiplicand:high_multiplicand,
                        each=high_multiplier-low_multiplier+1)
    multiplier <- rep(low_multiplier:high_multiplier,
                      high_multiplicand-low_multiplicand+1)
    product <- multiplicand * multiplier
    digits <- paste(as.character(multiplicand), as.character(multiplier),
                    as.character(product), sep="")

    df <- data.frame(multiplicand, multiplier, product, digits,
                     stringsAsFactors=FALSE)

    candidates <- product < 10000 & !grepl('0', multiplicand) &
        !grepl('0', multiplier) & !grepl('0', product)

    df <- df[candidates,]

    pandigital <- sapply(df$digits, function(i) {
        length(unique(as.integer(strsplit(i, split=NULL)[[1]]))) == 9
    })
    return(df$product[pandigital])
}

products <- vector()
products <- c(products, get_products(1, 9, 1234, 9876))
products <- c(products, get_products(12, 98, 123, 987))

cat(sum(unique(products)), fill=TRUE)

File: 32.js
#!/usr/bin/env node

// Pandigital products

var products = get_products(1, 9, 1234, 9876);
products = unique(products.concat(get_products(12, 98, 123, 987)));

var sum = products.reduce(function(a, b) {
    return a + b;
});

console.log(sum);

function get_products(lo_multiplicand, hi_multiplicand, lo_multiplier,
    hi_multiplier) {
    var products = [];

    for (var i = lo_multiplicand; i <= hi_multiplicand; i++) {
        if (i.toString().indexOf('0') >= 0) {
            continue;
        }
        for (var j = lo_multiplier; j <= hi_multiplier; j++) {
            if (j.toString().indexOf('0') >= 0) {
                continue;
            }
            var product = i * j;
            if (product >= 10000) {
                continue;
            }
            if (product.toString().indexOf('0') >= 0) {
                continue;
            }
            var digits = i.toString() + j.toString() + product.toString();
            digits = unique(digits.split(''));
            if (digits.length < 9) {
                continue;
            }
            products.push(product);
        }
    }

    return products;
}

function unique(array) {
    return array.reduce(function(prev, cur) {
        if (prev.indexOf(cur) < 0) {
            prev.push(cur);
        }
        return prev;
    }, []);
}
